modules/client-handler/src/main/java/org/apache/ignite/client/handler/ClientPrimaryReplicaTracker.java

```diff
@@ -0,0 +1,253 @@
+    }
+
+    long updateCount() {
+        return updateCount.get();
+    }
+
+    @SuppressWarnings({"rawtypes", "unchecked"})
+    void start() {
+        placementDriver.listen(PrimaryReplicaEvent.PRIMARY_REPLICA_ELECTED, (EventListener) this);
+        catalogService.listen(CatalogEvent.TABLE_DROP, (EventListener) this);
+    }
+
+    @SuppressWarnings({"rawtypes", "unchecked"})
+    void stop() {
+        catalogService.removeListener(CatalogEvent.TABLE_DROP, (EventListener) this);
+        placementDriver.removeListener(PrimaryReplicaEvent.PRIMARY_REPLICA_ELECTED, (EventListener) this);
+        primaryReplicas.clear();
+    }
+
+    private CompletableFuture<List<ReplicaHolder>> initReplicasForTableAsync(Integer tableId) {
+        try {
+            // Initially, request all primary replicas for the table.
+            // Then keep them updated via PRIMARY_REPLICA_ELECTED events.
+            long timestamp = clock.nowLong();
+            CatalogTableDescriptor tableDesc = catalogService.table(tableId, timestamp);
+
+            if (tableDesc == null) {
+                // TODO: Should we wait for table sync?
+                return CompletableFuture.failedFuture(tableNotFoundException(tableId));
+            }
+
+            CatalogZoneDescriptor zoneDesc = catalogService.zone(tableDesc.zoneId(), timestamp);
+
+            if (zoneDesc == null) {
+                return CompletableFuture.failedFuture(tableNotFoundException(tableId));
+            }
+
+            return primaryReplicasAsyncInternal(tableId, zoneDesc.partitions());
+        } catch (Throwable t) {
+            return CompletableFuture.failedFuture(t);
+        }
+    }
+
+    private CompletableFuture<List<ReplicaHolder>> primaryReplicasAsyncInternal(int tableId, int partitions) {
+        CompletableFuture<ReplicaMeta>[] futs = (CompletableFuture<ReplicaMeta>[]) new CompletableFuture[partitions];
+
+        for (int partition = 0; partition < partitions; partition++) {
+            futs[partition] = placementDriver.getPrimaryReplica(new TablePartitionId(tableId, partition), clock.now());
+        }
+
+        CompletableFuture<Void> all = CompletableFuture.allOf(futs);
+
+        return all.thenApply(v -> {
+            List<ReplicaHolder> replicaHolders = new ArrayList<>(partitions);
+
+            for (int partition = 0; partition < partitions; partition++) {
+                ReplicaMeta replicaMeta = futs[partition].join();
+
+                // Returning null is fine - the client will use default channel for this partition.
+                replicaHolders.add(replicaMeta == null
+                        ? new ReplicaHolder(null, HybridTimestamp.MIN_VALUE)
+                        : new ReplicaHolder(replicaMeta.getLeaseholder(), replicaMeta.getStartTime()));
+            }
+
+            return replicaHolders;
+        });
+    }
+
+    @Override
+    public CompletableFuture<Boolean> notify(EventParameters parameters, @Nullable Throwable exception) {
```

Comment:
Instead of having one notify for both catalog and replica events I'd rather parametrize each listener with corresponding handler. It's however a matter of taste - up to you.

Reply:
I'd prefer to keep it as is, since type erasure does not allow me to implement the same interface multiple times in the same class with different generic type arguments. The workaround is to put handlers into fields, but this seems like more code for no good reason.